### ISSUE TITLE
Improve SD ESLint rules.

### ISSUE THREAD

**Issue by wdstar** [screwdriver-cd/screwdriver#1650]
**What happened**:

Now SD JS codes have many big functions, which readability are very low and it's not easy to fix or improve it. 

In the current [eslint-config-screwdriver](https://github.com/screwdriver-cd/eslint-config-screwdriver), the following rules are not activated yet. We should activate those rules by `warn` level and refactor SD sources step by step.
- [max-lines-per-function](https://eslint.org/docs/rules/max-lines-per-function): default max: 50
- [complexity](https://eslint.org/docs/rules/complexity): default max: 20
- [max-params](https://eslint.org/docs/rules/max-params): default max: 3
- [max-statements](https://eslint.org/docs/rules/max-statements): default max: 10

**References**:

- Currently source codes violations (excludes `test/` directory):

|Rules|[screwdriver](https://github.com/screwdriver-cd/screwdriver)|[models](https://github.com/screwdriver-cd/models)|
|:---|---:|---:|
|max-lines-per-function|121|28|
|complexity|3|8|
|max-params|19|7|
|max-statements|34|21|

Please post any comments. (e.g. we should activate other rules,...)

**Comment by jithin1987** [comment]
cc @DekusDenial @adong

**Comment by DekusDenial** [comment]
@wdstar what do you say if we up `eslint-config-screwdriver` to `^4.0.0` and add `prettier` and following mostly https://github.com/screwdriver-cd/ui/blob/master/.eslintrc.js minus the ember part?

**Comment by wdstar** [comment]
@DekusDenial At first SD API's [current sources](https://github.com/screwdriver-cd/screwdriver) has new 453 errors if we update `eslint-config-screwdriver` to `^4.0.0`. 28 errors still remain after hitting `eslint --fix .` It takes some cost to fix its errors (mostly `max-len` violation) manually.

```
$ git diff
diff --git a/package.json b/package.json
index f424079..d1ecb52 100644
--- a/package.json
+++ b/package.json
@@ -123,7 +123,7 @@
     "coveralls": "^2.11.12",
     "cucumber": "2.3.1",
     "eslint": "^5.16.0",
-    "eslint-config-screwdriver": "^3.0.1",
+    "eslint-config-screwdriver": "^4.0.0",
     "form-data": "^2.3.3",
     "jenkins-mocha": "^7.0.0",
     "mockery": "^2.0.0",
$ npm install
...
$ eslint .
...
✖ 583 problems (453 errors, 130 warnings)
  446 errors and 0 warnings potentially fixable with the `--fix` option.

$ eslint --fix .
...
✖ 158 problems (28 errors, 130 warnings)
```

Ref.: Diff after eslint --fix
https://github.com/sonic-screwdriver-cd/screwdriver/commit/92bcf30a93405d04e7c6612f49f76312df7f9555

**Comment by wdstar** [comment]
And then we introduce `prettier`.
Basically we should  do formatting by `prettier` and do linting by `eslint`.

- Add `prettier` and its plugin for `eslint`. `eslint-config-prettier` deactivates eslint's formatting rules not to conflict prettier's formatting.
```
$ npm install eslint prettier eslint-config-prettier eslint-plugin-prettier -D
```
- Activate prettier's recommended formatting.
```diff
$ git diff .eslintrc.yaml
diff --git a/.eslintrc.yaml b/.eslintrc.yaml
index 6967af5..344c57c 100644
--- a/.eslintrc.yaml
+++ b/.eslintrc.yaml
@@ -1,6 +1,8 @@
-extends: screwdriver
+extends:
+    - "screwdriver"
+    - "plugin:prettier/recommended"  # last one wins
 rules:
     max-lines-per-function: "warn"
     complexity: "warn"
     max-params: "warn"
```
- Auto fixing... Only 5 linting errors still remain. We must fix them manually. 172 warnings by my first proposal rules.
```
$ eslint .
...
✖ 5144 problems (5014 errors, 130 warnings)
  5009 errors and 0 warnings potentially fixable with the `--fix` option.
$ eslint --fix .
...
✖ 177 problems (5 errors, 172 warnings)

$ eslint . | grep error
  29:24  error    Expected the Promise rejection reason to be an Error           prefer-promise-reject-errors
  50:29  error    Use array destructuring                                             prefer-destructuring
  117:21  error  new Buffer() is deprecated. Use Buffer.from(), Buffer.alloc(), or Buffer.allocUnsafe() instead  no-buffer-constructor
  44:25  error    Unexpected use of 'isFinite'                                             no-restricted-globals
  57:29  error    Use object destructuring                                                 prefer-destructuring
✖ 177 problems (5 errors, 172 warnings)
```
- Make sure of unit tests. All green!
```
$ jenkins-mocha --recursive --timeout 3000 --exit
...
  701 passing (5m)


=============================== Coverage summary ===============================
Statements   : 99.87% ( 3145/3149 )
Branches     : 98.38% ( 971/987 )
Functions    : 99.65% ( 843/846 )
Lines        : 99.93% ( 3010/3012 )
================================================================================
```

Ref.: Diff after introducing `prettier`
https://github.com/sonic-screwdriver-cd/screwdriver/commit/c71112e80e1bd245a6a0216f083b0af6116ae27b

In conclusion, we should use `prettier` to free from annoyance of source code formatting.
@DekusDenial Thank you for your advice.

**Comment by jithine** [comment]
We are now on latest `eslint` and latest `eslint-config-airbnb-base`
Screwdriver API is now using it, we will gradually move remaining libraries to ```"eslint-config-screwdriver": "^5"```
